// @blueprintjs/core/lib/esnext/components/html/html.js
/*
 * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the terms of the LICENSE file distributed with this project.
 */
import classNames from "classnames";
import * as React from "react";
import { BLOCKQUOTE, CODE, CODE_BLOCK, HEADING, LABEL, LIST } from "../../common/classes";
function htmlElement(tagName, tagClassName) {
    return props => {
        const { className, elementRef, ...htmlProps } = props;
        return React.createElement(tagName, {
            ...htmlProps,
            className: classNames(tagClassName, className),
            ref: elementRef,
        });
    };
}
// the following components are linted by blueprint-html-components because
// they should rarely be used without the Blueprint classes/styles:
export const H1 = htmlElement("h1", HEADING);
export const H2 = htmlElement("h2", HEADING);
export const H3 = htmlElement("h3", HEADING);
export const H4 = htmlElement("h4", HEADING);
export const H5 = htmlElement("h5", HEADING);
export const H6 = htmlElement("h6", HEADING);
export const Blockquote = htmlElement("blockquote", BLOCKQUOTE);
export const Code = htmlElement("code", CODE);
export const Pre = htmlElement("pre", CODE_BLOCK);
export const Label = htmlElement("label", LABEL);
// these two are not linted by blueprint-html-components because there are valid
// uses of these elements without Blueprint styles:
export const OL = htmlElement("ol", LIST);
export const UL = htmlElement("ul", LIST);
