// @blueprintjs/core/lib/esnext/components/text/text.js
/*
 * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the terms of the LICENSE file distributed with this project.
 */
import classNames from "classnames";
import * as React from "react";
import * as Classes from "../../common/classes";
import { DISPLAYNAME_PREFIX } from "../../common/props";
export class Text extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.state = {
            isContentOverflowing: false,
            textContent: "",
        };
        this.textRef = null;
    }
    componentDidMount() {
        this.update();
    }
    componentDidUpdate() {
        this.update();
    }
    render() {
        const classes = classNames({
            [Classes.TEXT_OVERFLOW_ELLIPSIS]: this.props.ellipsize,
        }, this.props.className);
        const { tagName: TagName = "div" } = this.props;
        return (React.createElement(TagName, { className: classes, ref: (ref) => (this.textRef = ref), title: this.state.isContentOverflowing ? this.state.textContent : undefined }, this.props.children));
    }
    update() {
        if (this.textRef == null) {
            return;
        }
        const newState = {
            isContentOverflowing: this.props.ellipsize && this.textRef.scrollWidth > this.textRef.clientWidth,
            textContent: this.textRef.textContent,
        };
        this.setState(newState);
    }
}
Text.displayName = `${DISPLAYNAME_PREFIX}.Text`;
